src/Crypto/Signing/HsmSignerService.cs

```diff
@@ -0,0 +1,178 @@
+// Copyright 2020 De Staat der Nederlanden, Ministerie van Volksgezondheid, Welzijn en Sport.
+// Licensed under the EUROPEAN UNION PUBLIC LICENCE v. 1.2
+// SPDX-License-Identifier: EUPL-1.2
+
+using System;
+using System.Collections.Generic;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Net.Http.Json;
+using System.Security.Cryptography;
+using System.Security.Cryptography.X509Certificates;
+using System.Text;
+using System.Threading.Tasks;
+using NL.Rijksoverheid.ExposureNotification.BackEnd.Domain;
+
+namespace NL.Rijksoverheid.ExposureNotification.BackEnd.Crypto.Signing
+{
+    /// <summary>
+    /// Typed HttpClient to request a signature from the HsmSignerService API
+    /// </summary>
+    public class HsmSignerService : IHsmSignerService
+    {
+        private readonly HttpClient _httpClient;
+        private readonly IHsmSignerConfig _config;
+
+        public HsmSignerService(
+            HttpClient httpClient,
+            IHsmSignerConfig config
+            )
+        {
+            _httpClient = httpClient ?? throw new ArgumentNullException(nameof(httpClient));
+            _config = config ?? throw new ArgumentNullException(nameof(config));
+
+            _httpClient.BaseAddress = new Uri(_config.BaseAddress);
+            _httpClient.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+        }
+
+        public async Task<byte[]> GetNlSignatureAsync(byte[] content)
+        {
+            using var sha256Hasher = SHA256.Create();
+            var contentHash = sha256Hasher.ComputeHash(content);
+            var contentHashBased64 = Convert.ToBase64String(contentHash);
+
+            var certificateChain = GetCertificateChain();
+            var nlPem = _config.NlPem;
+
+            var builder = new StringBuilder();
+
+            // Add EV RSA certificate
+            builder.AppendLine(nlPem);
+
+            // Add EV certificate chain
+            foreach (var cert in certificateChain)
+            {
+                builder.AppendLine(new string(PemEncoding.Write("CERTIFICATE", cert.RawData)));
+            }
+
+            var pem = builder.ToString();
+
+            var pemBytes = Encoding.UTF8.GetBytes(pem);
+            var pemBytesBased64 = Convert.ToBase64String(pemBytes);
+
+            var requestModel = new HsmSignerRequestModel()
+            {
+                Algorithm = "sha256",
+                Cert = pemBytesBased64,
+                Hash = contentHashBased64,
+                TimeStamp = true
+            };
+
+            _httpClient.DefaultRequestHeaders.Remove("Authorization");
+            _httpClient.DefaultRequestHeaders.Add("Authorization", _config.NlJwt);
+
+            var response = _httpClient.PostAsJsonAsync("/cms", requestModel).Result;
+
+            if (!response.IsSuccessStatusCode)
+            {
+                //TODO: error handling
+                return Array.Empty<byte>();
```

Comment:
Could we just throw here? To prevent invalid signatures?

Reply:
Yeah, good point. Throwing some semi-random exceptions for now, so we can get testing 😅